File: src/net/mm2d/wcc/SVSection.java
/**
 * Copyright(c) 2014 大前良介(OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.wcc;

import net.mm2d.color.ColorMethod;

import java.awt.Color;
import java.awt.Dimension;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.image.BufferedImage;

import javax.swing.JPanel;

/**
 * 輝度彩度を円柱モデル断面で操作するUI
 *
 * @author <a href="mailto:ryo@mm2d.net">大前良介(OHMAE Ryosuke)</a>
 */
public class SVSection extends JPanel {
    private static final int RANGE = 255;
    private static final int WIDTH = 260;
    private static final int HEIGHT = 260;
    private final BufferedImage mImage;
    private final int mMarginTop;
    private final int mMarginLeft;
    private float mHue;
    private float mSaturation;
    private float mValue;
    private OnSVChangeListener mListener;

    /**
     * 値変化のリスナー
     */
    public interface OnSVChangeListener {
        /**
         * 値が変化したときコール
         *
         * @param hue 色相
         * @param saturation 彩度
         * @param value 輝度
         */
        public void onSVChange(float hue, float saturation, float value);
    }

    /**
     * インスタンス作成
     */
    public SVSection() {
        super();
        mHue = 0.0f;
        mSaturation = 1.0f;
        mValue = 1.0f;
        mImage = new BufferedImage(RANGE, RANGE, BufferedImage.TYPE_4BYTE_ABGR);
        mMarginTop = (WIDTH - RANGE) / 2;
        mMarginLeft = (HEIGHT - RANGE) / 2;
        makeSVSection(mImage);
        setPreferredSize(new Dimension(WIDTH, HEIGHT));
        final MouseAdapter adapter = new MouseAdapter() {
            @Override
            public void mousePressed(MouseEvent e) {
                selectPoint(e.getX(), e.getY());
            }

            @Override
            public void mouseDragged(MouseEvent e) {
                selectPoint(e.getX(), e.getY());
            }
        };
        addMouseListener(adapter);
        addMouseMotionListener(adapter);
    }

    /**
     * マウス操作による値の選択
     *
     * @param x X座標
     * @param y Y座標
     */
    private void selectPoint(int x, int y) {
        final float s = ColorMethod.toFloat(ColorMethod.clamp(x - mMarginLeft, 0, 255));
        final float v = ColorMethod.toFloat(ColorMethod.clamp(RANGE - (y - mMarginTop), 0, 255));
        setHSV(mHue, s, v, true);
        repaint();
    }

    /**
     * HSVの値を設定する
     *
     * @param h H
     * @param s S
     * @param v V
     */
    public void setHSV(float h, float s, float v) {
        setHSV(h, s, v, false);
    }

    /**
     * HSVの値を設定する
     *
     * @param h H
     * @param s S
     * @param v V
     * @param notify リスナー通知の有無
     */
    private void setHSV(float h, float s, float v, boolean notify) {
        if (mHue != h) {
            mHue = h;
            makeSVSection(mImage);
        }
        mSaturation = s;
        mValue = v;
        if (notify) {
            performSVChange();
        }
        repaint();
    }

    /**
     * SV断面の画像を作成する
     *
     * @param image 書き込み先
     */
    private void makeSVSection(BufferedImage image) {
        for (int y = 0; y < image.getHeight(); y++) {
            for (int x = 0; x < image.getWidth(); x++) {
                final float[] rgb = ColorMethod.HSVtoRGB(
                        mHue,
                        (float) x / RANGE,
                        (float) (RANGE - y) / RANGE);
                final int color = ColorMethod.toColor(rgb);
                image.setRGB(x, y, color);
            }
        }
    }

    /**
     * 値変化のリスナーを登録する
     *
     * @param listener リスナー
     */
    public void setOnSVChangeListener(OnSVChangeListener listener) {
        mListener = listener;
    }

    /**
     * 値の変化をリスナーに通知する
     */
    private void performSVChange() {
        if (mListener != null) {
            mListener.onSVChange(mHue, mSaturation, mValue);
        }
    }

    @Override
    public void paint(Graphics g) {
        final Graphics2D g2 = (Graphics2D) g;
        g2.setBackground(getBackground());
        g2.clearRect(0, 0, getWidth(), getHeight());
        g2.drawImage(mImage, mMarginLeft, mMarginTop, this);
        // 選択している点を描画
        final int x = (int) (mSaturation * RANGE) + mMarginLeft;
        final int y = (int) (RANGE - mValue * RANGE) + mMarginTop;
        g2.setXORMode(Color.WHITE);
        g2.drawRect(x - 1, y - 1, 2, 2);
    }
}

File: src/main/java/net/mm2d/wcc/SvSection.kt
/*
 * Copyright (c) 2014 大前良介 (OHMAE Ryosuke)
 *
 * This software is released under the MIT License.
 * http://opensource.org/licenses/MIT
 */

package net.mm2d.wcc

import net.mm2d.color.ColorUtils
import net.mm2d.color.clamp
import net.mm2d.color.toRatio

import java.awt.Color
import java.awt.Dimension
import java.awt.Graphics
import java.awt.Graphics2D
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.awt.image.BufferedImage

import javax.swing.JPanel

/**
 * 輝度彩度を円柱モデル断面で操作するUI
 */
class SvSection : JPanel() {
    private val image = BufferedImage(RANGE + 1, RANGE + 1, BufferedImage.TYPE_4BYTE_ABGR)
    private val marginTop: Int = (WIDTH - RANGE) / 2
    private val marginLeft: Int = (HEIGHT - RANGE) / 2
    private var hue: Float = 0f
    private var saturation: Float = 1f
    private var value: Float = 1f
    private var maxColor: Color = Color.RED
    var onSvChangeListener: ((hue: Float, saturation: Float, value: Float) -> Unit)? = null

    /**
     * インスタンス作成
     */
    init {
        for (y in 0..RANGE) {
            for (x in 0..RANGE) {
                val s = x.toFloat() / RANGE
                val v = (RANGE - y).toFloat() / RANGE
                image.setRGB(x, y, ColorUtils.svToMask(s, v))
            }
        }
        preferredSize = Dimension(WIDTH, HEIGHT)
        addMouseListener(object : MouseAdapter() {
            override fun mousePressed(e: MouseEvent) {
                selectPoint(e.x, e.y)
            }
        })
        addMouseMotionListener(object : MouseAdapter() {
            override fun mouseDragged(e: MouseEvent) {
                selectPoint(e.x, e.y)
            }
        })
    }

    /**
     * マウス操作による値の選択
     *
     * @param x X座標
     * @param y Y座標
     */
    private fun selectPoint(x: Int, y: Int) {
        val s = (x - marginLeft).clamp(0, 255).toRatio()
        val v = (RANGE - (y - marginTop)).clamp(0, 255).toRatio()
        setHsv(hue, s, v, true)
    }

    /**
     * HSVの値を設定する
     *
     * @param h H
     * @param s S
     * @param v V
     */
    fun setHsv(h: Float, s: Float, v: Float) {
        setHsv(h, s, v, false)
    }

    /**
     * HSVの値を設定する
     *
     * @param h H
     * @param s S
     * @param v V
     * @param notify リスナー通知する場合true
     */
    private fun setHsv(h: Float, s: Float, v: Float, notify: Boolean) {
        if (hue != h) {
            hue = h
            maxColor = Color(ColorUtils.hsvToColor(hue, 1f, 1f))
        }
        saturation = s
        value = v
        if (notify) {
            onSvChangeListener?.invoke(h, s, v)
        }
        repaint()
    }


    override fun paint(g: Graphics?) {
        val g2 = g as? Graphics2D ?: return
        g2.background = background
        g2.clearRect(0, 0, width, height)
        g2.color = maxColor
        g2.fillRect(marginLeft, marginTop, image.width, image.height)
        g2.drawImage(image, marginLeft, marginTop, this)
        // 選択している点を描画
        val x = (saturation * RANGE).toInt() + marginLeft
        val y = (RANGE - value * RANGE).toInt() + marginTop
        g2.setXORMode(Color.WHITE)
        g2.drawRect(x - 1, y - 1, 2, 2)
    }

    companion object {
        private const val RANGE = 255
        private const val WIDTH = 260
        private const val HEIGHT = 260
    }
}
